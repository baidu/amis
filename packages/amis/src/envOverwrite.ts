/**
 * @file 用于在移动端或不同语言环境下使用不同配置
 */

import {SchemaNode, Schema} from './types';
import {RendererProps, RendererConfig, addSchemaFilter} from './factory';
import {findObjectsWithKey} from './utils/helper';
const isMobile = (window as any).matchMedia?.('(max-width: 768px)').matches
  ? true
  : false;

// 这里不能用 addSchemaFilter 是因为还需要更深层的替换，比如 select 里的 options
export const envOverwrite = (schema: any, locale?: string) => {
  if (schema.mobile && isMobile) {
    Object.assign(schema, schema.mobile);
    delete schema.mobile;
  }

  if (locale) {
    let schemaNodes = findObjectsWithKey(schema, locale);
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode[locale]);
      delete schemaNode[locale];
    }
  }

  if (isMobile) {
    let schemaNodes = findObjectsWithKey(schema, 'mobile');
    for (let schemaNode of schemaNodes) {
      Object.assign(schemaNode, schemaNode['mobile']);
      delete schemaNode['mobile'];
    }
  }
};
