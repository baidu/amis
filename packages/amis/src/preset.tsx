import {
  addRootWrapper,
  extendDefaultEnv,
  LazyComponent,
  render,
  themeable,
  ThemeProps
} from 'amis-core';

import {alert, confirm, setRenderSchemaFn, toast, ImageGallery} from 'amis-ui';
import React from 'react';

extendDefaultEnv({
  alert,
  confirm,
  notify: (type: keyof typeof toast, msg: string, conf: any) =>
    toast[type] ? toast[type](msg, conf) : console.warn('[Notify]', type, msg)
});

setRenderSchemaFn((controls, value, callback, scopeRef, theme) => {
  return render(
    {
      name: 'form',
      type: 'form',
      wrapWithPanel: false,
      mode: 'horizontal',
      controls,
      messages: {
        validateFailed: ''
      }
    },
    {
      data: value,
      onFinished: callback,
      scopeRef,
      theme
    },
    {
      session: 'prompt'
    }
  );
});

addRootWrapper((props: any) => {
  const {env, children} = props;
  return (
    <ImageGallery modalContainer={env.getModalContainer}>
      {children}
    </ImageGallery>
  );
});

const SimpleSpinner = themeable(
  (
    props: {
      className?: string;
      spinnerClassName?: string;
    } & ThemeProps
  ) => {
    const cx = props.classnames;

    return (
      <div
        data-testid="spinner"
        className={cx(`Spinner`, 'in', props.className)}
      >
        <div
          className={cx(
            `Spinner-icon`,
            'Spinner-icon--default',
            props.spinnerClassName
          )}
        ></div>
      </div>
    );
  }
);

LazyComponent.defaultProps.placeholder = <SimpleSpinner />;
