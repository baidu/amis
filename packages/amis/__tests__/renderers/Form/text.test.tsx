import React = require('react');
import {render, cleanup, fireEvent, waitFor} from '@testing-library/react';
import '../../../src';
import {render as amisRender} from '../../../src';
import {makeEnv, replaceReactAriaIds, wait} from '../../helper';
import {clearStoresCache} from '../../../src';

afterEach(() => {
  cleanup();
  clearStoresCache();
});

const setup = async (
  inputOptions: any = {},
  formOptions: any = {},
  formItems: any[] = [{}]
) => {
  const utils = render(
    amisRender(
      {
        type: 'form',
        api: '/api/mock2/form/saveForm',
        body: [
          {
            name: 'text',
            label: 'text',
            type: 'input-text',
            changeImmediately: true,
            ...inputOptions
          },
          ...formItems
        ],
        ...formOptions
      },
      {},
      makeEnv()
    )
  );

  await waitFor(() => {
    expect(
      utils.container.querySelector('input[name="text"]')
    ).toBeInTheDocument();

    expect(
      utils.container.querySelector('button[type="submit"]')
    ).toBeInTheDocument();
  });

  const input = utils.container.querySelector(
    'input[name="text"]'
  ) as HTMLInputElement;

  const submitBtn = utils.container.querySelector(
    'button[type="submit"]'
  ) as HTMLElement;

  return {
    input,
    submitBtn,
    ...utils
  };
};

/**
 * 基本使用
 */
test('Renderer:text', async () => {
  const {container, input} = await setup();

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
  // 输入是否正常
  fireEvent.change(input, {target: {value: 'AbCd'}});
  // 事件机制导致hanleChange变为异步
  await wait(300);
  expect(input.value).toBe('AbCd');
});

/**
 * type 为 url，主要测试校验
 */
test('Renderer:text type is url', async () => {
  const {container, input, submitBtn} = await setup({
    type: 'input-url'
  });

  fireEvent.change(input, {target: {value: 'abcd'}});
  fireEvent.click(submitBtn);
  await wait(200); // 表单校验是异步的，所以必须要等一段时间 @todo 感觉可能需要寻找更靠谱的办法

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('validate fail');

  fireEvent.change(input, {target: {value: 'https://www.baidu.com'}});
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('validate success');
});

/**
 * type 为 email，主要测试校验
 */
test('Renderer:text type is email', async () => {
  const {container, input, submitBtn} = await setup({
    type: 'input-email'
  });

  fireEvent.change(input, {target: {value: 'abcd'}});
  fireEvent.click(submitBtn);
  await wait(200);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('validate fail');

  fireEvent.change(input, {target: {value: 'test@baidu.com'}});
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('validate success');
});

/**
 * type 为 password
 */
test('Renderer:text type is password', async () => {
  const {container, input} = await setup({
    type: 'input-password'
  });

  fireEvent.change(input, {target: {value: 'abcd'}});
  await wait(300);

  expect(input.getAttribute('type')).toBe('password');

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('password invisible');

  const revealPasswordBtn = container.querySelector(
    '.cxd-TextControl-revealPassword'
  ) as HTMLElement;

  fireEvent.click(revealPasswordBtn);

  await wait(300);

  expect(input.getAttribute('type')).toBe('text');

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('password visible');
});

/**
 * type 为 password revealPassword
 */
test('Renderer:text type is password with revealPassword', async () => {
  const {container, input} = await setup({
    type: 'input-password',
    revealPassword: false
  });

  expect(
    container.querySelector('.cxd-TextControl-revealPassword') as Element
  ).not.toBeInTheDocument();

  fireEvent.change(input, {target: {value: 'abcd'}});
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
});

/**
 * 配置addOn
 */
test('Renderer:text with addOn', async () => {
  const {container, input} = await setup({
    addOn: {
      type: 'button',
      label: '搜索'
    }
  });

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
});

/**
 * 配置 clearable
 */
test('Renderer:text with clearable', async () => {
  const {container, input} = await setup({
    clearable: true
  });
  fireEvent.change(input, {target: {value: 'abcd'}}); // 有值之后才会显示clear的icon
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();

  fireEvent.click(
    container.querySelector('a.cxd-TextControl-clear') as HTMLElement
  );
  await wait(300);
  expect(input.value).toBe('');
});

/**
 * 选择器模式
 */
test('Renderer:text with options', async () => {
  const {container, input} = await setup({
    options: [
      {
        label: 'Option A',
        value: 'a'
      },
      {
        label: 'Option B',
        value: 'b'
      }
    ]
  });

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();

  // 展开 options
  fireEvent.click(
    container.querySelector('.cxd-TextControl-input') as HTMLElement
  );
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('options is open');

  // 选中一项
  fireEvent.click(
    container.querySelector(
      '.cxd-TextControl-sugs .cxd-TextControl-sugItem'
    ) as HTMLElement
  );
  await wait(300);
  // expect(input.value).toBe('a');

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('select first option');
});

/**
 * 选择器模式，多选、分隔符、提取值
 */
test('Renderer:text with options and multiple and delimiter', async () => {
  const {container, input} = await setup(
    {
      multiple: true,
      options: [
        {
          label: 'OptionA',
          value: 'a'
        },
        {
          label: 'OptionB',
          value: 'b'
        },
        {
          label: 'OptionC',
          value: 'c'
        },
        {
          label: 'OptionD',
          value: 'd'
        }
      ],
      delimiter: '-',
      joinValues: true,
      creatable: true
    },
    {},
    [
      {
        type: 'static',
        name: 'text'
      }
    ]
  );

  const textControl = container.querySelector(
    '.cxd-TextControl-input'
  ) as HTMLElement;

  // 展开 options
  fireEvent.click(textControl);
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('options is opened');

  // 选中第一项
  fireEvent.click(
    container.querySelector(
      '.cxd-TextControl-sugs .cxd-TextControl-sugItem'
    ) as HTMLElement
  );
  await wait(300);
  // expect(input.value).toBe('a');

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('first option selected');

  // 再次打开 options
  fireEvent.click(textControl);
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot(
    'options is opened again, and first option already selected'
  );

  // 选中 options 中的第一项
  fireEvent.click(
    container.querySelector(
      '.cxd-TextControl-sugs .cxd-TextControl-sugItem'
    ) as HTMLElement
  );
  await wait(300);

  // 分隔符
  expect(
    (container.querySelector('.cxd-PlainField') as Element).innerHTML
  ).toBe('a-b');

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('second option selected');

  // 可创建
  fireEvent.click(textControl);
  await wait(300);
  fireEvent.change(input, {target: {value: 'AbCd'}});
  await wait(500);
  fireEvent.keyDown(input, {key: 'Enter', code: 13});
  await wait(500);

  expect(
    (container.querySelector('.cxd-PlainField') as Element).innerHTML
  ).toBe('a-b-AbCd');

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot('thrid option create');
});

/**
 * 前缀和后缀
 */
test('Renderer:text with prefix and suffix', async () => {
  const {container} = await setup({
    prefix: '￥',
    suffix: 'RMB'
  });

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
});

/**
 * 显示计数器
 */
test('Renderer:text with counter', async () => {
  const {container, input} = await setup({
    showCounter: true
  });

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();

  fireEvent.change(input, {target: {value: 'abcd'}});
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
});

/**
 * 显示计数器且配置最大值
 */
test('Renderer:text with counter and maxLength', async () => {
  const {container, input} = await setup({
    showCounter: true,
    maxLength: 10
  });

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();

  fireEvent.change(input, {target: {value: 'abcd'}});
  await waitFor(() => {
    expect(
      container.querySelector('[name="text"][value="abcd"]')
    ).toBeInTheDocument();
  });
  await wait(300);

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
});

/**
 * 转小写
 */
test('Renderer:text with transform lowerCase', async () => {
  const {input} = await setup({transform: {lowerCase: true}});

  fireEvent.change(input, {target: {value: 'AbCd'}});
  await wait(300);
  expect(input.value).toBe('abcd');
});

/**
 * 转大写
 */
test('Renderer:text with transform upperCase', async () => {
  const {input} = await setup({transform: {upperCase: true}});

  fireEvent.change(input, {target: {value: 'AbCd'}});
  await wait(300);
  expect(input.value).toBe('ABCD');
});

/**
 * 配置 resetValue and trimContents
 */
test('Renderer:text with resetValue and trimContents', async () => {
  const {container, input, submitBtn} = await setup(
    {
      resetValue: 'reset-value',
      value: 'text-value',
      trimContents: true
    },
    {},
    [
      {
        type: 'action',
        actionType: 'reset',
        target: 'text',
        className: 'reset-button'
      }
    ]
  );

  const resetBtn = container.querySelector('.cxd-Button.reset-button');
  await waitFor(() => {
    expect(resetBtn).toBeInTheDocument();
  });
  fireEvent.click(resetBtn!);

  await wait(500);

  expect(input.value).toBe('reset-value');

  // trimContents
  const textControl = container.querySelector(
    '.cxd-TextControl-input'
  ) as HTMLElement;

  fireEvent.click(textControl);
  await wait(300);
  fireEvent.change(input, {target: {value: '  abcde  '}});
  await wait(500);
  fireEvent.blur(input);
  await wait(500);

  expect(input.value).toBe('abcde');
});

/**
 * 配置 minLength、borderMode and className
 */
test('Renderer:text with minLength', async () => {
  const {container, input, submitBtn} = await setup(
    {
      minLength: 5,
      maxLength: 8,
      borderMode: 'half',
      inputControlClassName: 'test-text-class-one',
      nativeInputClassName: 'test-text-class-two'
    },
    {}
  );

  const textControl = container.querySelector(
    '.cxd-TextControl-input'
  ) as HTMLElement;

  // 测试 minLength
  fireEvent.click(textControl);
  await wait(300);
  fireEvent.change(input, {target: {value: '1234'}});
  await wait(500);
  fireEvent.click(submitBtn);
  await wait(300);
  expect(
    container.querySelector('.cxd-TextControl.has-error--minLength') as Element
  ).toBeInTheDocument();

  // 测试 maxLength
  fireEvent.click(textControl);
  await wait(300);
  fireEvent.change(input, {target: {value: '123456789'}});
  await wait(500);
  expect(
    container.querySelector('.cxd-TextControl.has-error--maxLength') as Element
  ).toBeInTheDocument();

  replaceReactAriaIds(container);
  expect(container).toMatchSnapshot();
});
