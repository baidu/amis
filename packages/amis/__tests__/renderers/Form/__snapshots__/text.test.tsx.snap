// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer:text 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                name="text"
                placeholder=""
                size="10"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text type is password with revealPassword 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                name="text"
                placeholder=""
                size="10"
                type="password"
                value="abcd"
              />
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with addOn 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl cxd-TextControl--withAddOn"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                name="text"
                placeholder=""
                size="10"
                type="text"
                value=""
              />
            </div>
            <div
              class="cxd-TextControl-button"
            >
              <button
                class="cxd-Button cxd-Button--default cxd-Button--size-default"
                type="button"
              >
                <span>
                  搜索
                </span>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with clearable 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                name="text"
                placeholder=""
                size="10"
                type="text"
                value="abcd"
              />
              <a
                class="cxd-TextControl-clear"
              >
                <icon-mock
                  classname="icon icon-input-clear"
                  icon="input-clear"
                />
              </a>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with counter 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                name="text"
                placeholder=""
                size="10"
                type="text"
                value=""
              />
              <span
                class="cxd-TextControl-counter"
              >
                0
              </span>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with counter 2`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                name="text"
                placeholder=""
                size="10"
                type="text"
                value="abcd"
              />
              <span
                class="cxd-TextControl-counter"
              >
                4
              </span>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with counter and maxLength 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                maxlength="10"
                name="text"
                placeholder=""
                size="10"
                type="text"
                value=""
              />
              <span
                class="cxd-TextControl-counter"
              >
                0/10
              </span>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with counter and maxLength 2`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <input
                autocomplete="off"
                class=""
                maxlength="10"
                name="text"
                placeholder=""
                size="10"
                type="text"
                value="abcd"
              />
              <span
                class="cxd-TextControl-counter"
              >
                4/10
              </span>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with minLength 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal is-error has-error--maxLength"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control is-error has-error--maxLength cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input cxd-TextControl-input--borderHalf test-text-class-one"
            >
              <input
                autocomplete="off"
                class="test-text-class-two"
                maxlength="8"
                minlength="5"
                name="text"
                placeholder=""
                size="10"
                type="text"
                value="123456789"
              />
            </div>
          </div>
          <ul
            class="cxd-Form-feedback"
          >
            <li>
              请控制内容长度, 不要输入 8 个以上字符
            </li>
          </ul>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with options 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              class="cxd-TextControl-input cxd-TextControl-input--withAC"
              role="combobox"
            >
              <input
                aria-autocomplete="list"
                autocomplete="off"
                class=""
                id="downshift-0-input"
                name="text"
                size="10"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renderer:text with prefix and suffix 1`] = `
<div>
  <div
    class="cxd-Panel cxd-Panel--default cxd-Panel--form"
    data-role="container"
  >
    <div
      class="cxd-Panel-heading"
    >
      <h3
        class="cxd-Panel-title"
      >
        <span
          class="cxd-TplField fr-view"
        >
          <span>
            表单
          </span>
        </span>
      </h3>
    </div>
    <div
      class="cxd-Panel-body"
    >
      <form
        class="cxd-Form cxd-Form--normal"
        novalidate=""
      >
        <input
          style="display: none;"
          type="submit"
        />
        <div
          class="cxd-Form-item cxd-Form-item--normal"
          data-amis-name="text"
          data-role="form-item"
        >
          <label
            class="cxd-Form-label"
          >
            <span>
              <span
                class="cxd-TplField fr-view"
              >
                <span>
                  text
                </span>
              </span>
            </span>
          </label>
          <div
            class="cxd-Form-control cxd-TextControl"
          >
            <div
              class="cxd-TextControl-input"
            >
              <span
                class="cxd-TextControl-inputPrefix"
              >
                ￥
              </span>
              <input
                autocomplete="off"
                class=""
                name="text"
                placeholder=""
                size="10"
                type="text"
                value=""
              />
              <span
                class="cxd-TextControl-inputSuffix"
              >
                RMB
              </span>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="cxd-Panel-footerWrap"
    >
      <div
        class="cxd-Panel-btnToolbar cxd-Panel-footer"
      >
        <button
          class="cxd-Button cxd-Button--primary cxd-Button--size-default"
          type="submit"
        >
          <span>
            提交
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
